In [1]:
import pandas as pd
import json
import plotly.express as px
import plotly
import ast
from collections import Counter
import numpy as np

## Analysis of Gene Terms in DGIdb from 2023
Reanalysis of the log data obtained from DGIdb user search patterns in Q1/Q2 2023. This is intended to extract the gene terms most often searched from each search type and give them over to Anastasia who will subsequently determine the frequency of searches that have gene collisions.

### Load Data

In [2]:
df = pd.read_excel("../input/log_data.xlsx")
df[0:5]

Unnamed: 0                        time request_num  \
0           0  2023-05-03T11:45:03.737594      #95027   
1           0  2023-05-03T11:45:06.699989      #95027   
2           0  2023-05-03T11:45:10.303598      #94988   
3           0  2023-05-03T11:45:12.311825      #94988   
4           0  2023-05-03T11:45:15.174727      #94988   

                                   uuid method              path format  \
0  34b514ef-ea79-498b-83a8-ec5abff01760    GET  /drug_names.json   json   
1  47c0fa4a-7068-42ce-91d8-525f987568ff    GET      /genes/ICAM3   html   
2  1fc0e9d3-f6a2-4b00-baba-0fd9ce9b5405    GET        /genes/HTT   html   
3  205878ef-9bea-47c4-9cff-222a93839525    GET  /gene_names.json   json   
4  15d9ec98-d825-49ef-ad58-0fd398ea6839    GET  /drug_names.json   json   

        controller action  status  duration      view      db  \
0  DrugsController  names     200     58.31      0.14    0.00   
1  GenesController   show     200    895.10    755.20  129.73   
2  GenesController   show     200  32723.75  31973.05  739.82   
3  GenesController  names     200    151.56      0.12    0.00   
4  DrugsController  names     200     54.78      0.10    0.00   

              params location error  
0                 {}      NaN   NaN  
1  {'name': 'ICAM3'}      NaN   NaN  
2    {'name': 'HTT'}      NaN   NaN  
3                 {}      NaN   NaN  
4                 {}      NaN   NaN

### Clean Data

We are interested in the api/v2/interactions queries

In [3]:
df["path"].value_counts()[0:10]

path
/api/v2/interactions.json         25280
/                                 12654
/gene_names.json                   7051
/interaction_search_results        5748
/search_interactions               5241
/api/v2/interaction_types.json     3202
/drug_names.json                   2070
/search_categories                 1523
/downloads                         1064
/categories_search_results         1048
Name: count, dtype: int64

In [4]:
data = (
    df[df["path"] == "/api/v2/interactions.json"]
    .drop(labels="Unnamed: 0", axis=1)
    .reset_index(drop=True)
)
data[0:5]

time request_num  \
0  2023-05-03T11:59:21.273789      #95027   
1  2023-05-03T12:15:13.185862      #95027   
2  2023-05-03T12:25:02.801847      #95027   
3  2023-05-03T12:29:09.147971      #95027   
4  2023-05-03T12:29:51.541701      #95027   

                                   uuid method                       path  \
0  5a1bd777-14c2-44b1-87b8-4bcec01433c4    GET  /api/v2/interactions.json   
1  66a47e2f-c94d-451c-8cdb-c1a29617b3cb    GET  /api/v2/interactions.json   
2  9f3932bb-5611-4317-ab44-fab33992e283    GET  /api/v2/interactions.json   
3  3a24f3df-c114-4739-8bdc-f89203ec31a0    GET  /api/v2/interactions.json   
4  e88c50f0-2b4f-4c7f-ab0d-075095bce6c2    GET  /api/v2/interactions.json   

  format            controller        action  status  duration    view     db  \
0   json  ServicesV2Controller  interactions     200     46.52    7.66  21.43   
1   json  ServicesV2Controller  interactions     200    993.19  917.98  61.50   
2   json  ServicesV2Controller  interactions     200     37.76    3.25  23.25   
3   json  ServicesV2Controller  interactions     200     35.21    0.52  19.20   
4   json  ServicesV2Controller  interactions     200     34.51    3.20  20.37   

                   params location error  
0       {'genes': 'ATN1'}      NaN   NaN  
1  {'drugs': 'Scriptaid'}      NaN   NaN  
2      {'genes': 'GPR26'}      NaN   NaN  
3  {'drugs': 'Isoprenol'}      NaN   NaN  
4     {'genes': 'IL10RA'}      NaN   NaN

We want to pull out the genes from the params of the query data

In [5]:
def get_search_type(record):
    try:
        record = record.replace("'", '"')
        record = json.loads(record)
        keys = list(record.keys())
        if len(keys) > 0:
            return keys[0]
        else:
            return "KeyLength"
    except json.JSONDecodeError:
        return "JSONDecode"

In [6]:
data["params"][0]
test = data["params"][0].replace("'", '"')
test = json.loads(test)
list(test.keys())

['genes']

In [7]:
data["type"] = None
data["type"] = data["params"].apply(get_search_type)

Distribution of Types of Interaction Query sent (drugs vs genes)

In [8]:
data["type"].value_counts()

type
genes         23070
drugs          1647
KeyLength       543
JSONDecode       20
Name: count, dtype: int64

In [9]:
data = data[data["type"] == "genes"].reset_index(drop=True)
data[0:5]

time request_num  \
0  2023-05-03T11:59:21.273789      #95027   
1  2023-05-03T12:25:02.801847      #95027   
2  2023-05-03T12:29:51.541701      #95027   
3  2023-05-03T12:29:57.144888      #95027   
4  2023-05-03T12:30:01.291949      #95027   

                                   uuid method                       path  \
0  5a1bd777-14c2-44b1-87b8-4bcec01433c4    GET  /api/v2/interactions.json   
1  9f3932bb-5611-4317-ab44-fab33992e283    GET  /api/v2/interactions.json   
2  e88c50f0-2b4f-4c7f-ab0d-075095bce6c2    GET  /api/v2/interactions.json   
3  26747924-e565-43ac-b36e-9d802d4ee5c8    GET  /api/v2/interactions.json   
4  8c9d7628-ce57-4c3d-b74e-a7acd042cb6b    GET  /api/v2/interactions.json   

  format            controller        action  status  duration    view     db  \
0   json  ServicesV2Controller  interactions     200     46.52    7.66  21.43   
1   json  ServicesV2Controller  interactions     200     37.76    3.25  23.25   
2   json  ServicesV2Controller  interactions     200     34.51    3.20  20.37   
3   json  ServicesV2Controller  interactions     200     33.73    3.04  20.15   
4   json  ServicesV2Controller  interactions     200    189.23  115.09  61.95   

                params location error   type  
0    {'genes': 'ATN1'}      NaN   NaN  genes  
1   {'genes': 'GPR26'}      NaN   NaN  genes  
2  {'genes': 'IL10RA'}      NaN   NaN  genes  
3  {'genes': 'VSTM2A'}      NaN   NaN  genes  
4   {'genes': 'SH2B3'}      NaN   NaN  genes

In [10]:
def get_genes(record):
    try:
        record = record.replace("'", '"')
        record = json.loads(record)
        if record["genes"]:
            genes = record["genes"].split(",")
            return genes
        else:
            return "KeyLength"
    except:  
        pass

In [11]:
data["genes"] = None
data["genes"] = data["params"].apply(get_genes)
data[0:5]

time request_num  \
0  2023-05-03T11:59:21.273789      #95027   
1  2023-05-03T12:25:02.801847      #95027   
2  2023-05-03T12:29:51.541701      #95027   
3  2023-05-03T12:29:57.144888      #95027   
4  2023-05-03T12:30:01.291949      #95027   

                                   uuid method                       path  \
0  5a1bd777-14c2-44b1-87b8-4bcec01433c4    GET  /api/v2/interactions.json   
1  9f3932bb-5611-4317-ab44-fab33992e283    GET  /api/v2/interactions.json   
2  e88c50f0-2b4f-4c7f-ab0d-075095bce6c2    GET  /api/v2/interactions.json   
3  26747924-e565-43ac-b36e-9d802d4ee5c8    GET  /api/v2/interactions.json   
4  8c9d7628-ce57-4c3d-b74e-a7acd042cb6b    GET  /api/v2/interactions.json   

  format            controller        action  status  duration    view     db  \
0   json  ServicesV2Controller  interactions     200     46.52    7.66  21.43   
1   json  ServicesV2Controller  interactions     200     37.76    3.25  23.25   
2   json  ServicesV2Controller  interactions     200     34.51    3.20  20.37   
3   json  ServicesV2Controller  interactions     200     33.73    3.04  20.15   
4   json  ServicesV2Controller  interactions     200    189.23  115.09  61.95   

                params location error   type     genes  
0    {'genes': 'ATN1'}      NaN   NaN  genes    [ATN1]  
1   {'genes': 'GPR26'}      NaN   NaN  genes   [GPR26]  
2  {'genes': 'IL10RA'}      NaN   NaN  genes  [IL10RA]  
3  {'genes': 'VSTM2A'}      NaN   NaN  genes  [VSTM2A]  
4   {'genes': 'SH2B3'}      NaN   NaN  genes   [SH2B3]

### How many terms idenitfied as gene symbols were queried across all queries? (queries may have more than one gene term)

In [12]:
def flatten_comprehension(matrix):
    return [item for row in matrix for item in row]

In [13]:
queried_genes_list = data["genes"].tolist()

In [14]:
queried_genes_list = flatten_comprehension(queried_genes_list)

In [15]:
num_queried_gene_terms = len(queried_genes_list)
num_queried_gene_terms

188832

### How many interaction queries are requested using genes?

In [16]:
num_interaction_queries_using_genes = len(data["genes"])
num_interaction_queries_using_genes

23070

### How many unique gene symbols were queried between the 23,070 queries and 188,832 total symbols?

In [17]:
data["genes"] = data["genes"].apply(set)
data["genes"]

0                                         {ATN1}
1                                        {GPR26}
2                                       {IL10RA}
3                                       {VSTM2A}
4                                        {SH2B3}
                          ...                   
23065       {RBMY1C, RN7SL5P, PVR, QSOX1, CLIP1}
23066    {MFSD2A, GINM1, ZNF341, SNORD15B, OSR2}
23067       {VASP, POU4F1, SLPI, MKRN4P, TFAP2B}
23068       {MTCP1, HSD3B2, PTHLH, ITGAM, IDDM9}
23069           {JAM3, HPDL, SRRM4, MOB2, SLIRP}
Name: genes, Length: 23070, dtype: object

In [18]:
queried_gene_symbols_set = data["genes"].agg(lambda x: set.union(*x))
num_unique_gene_symbols_queried = len(queried_gene_symbols_set)
num_unique_gene_symbols_queried

43400

Count each queried gene

In [19]:
all_genes = [gene for sublist in data["genes"] for gene in sublist]
gene_counts = Counter(all_genes)
gene_counts

Counter({'XYZA': 1549,
         'TNF': 1001,
         'AP1': 715,
         'AP2': 576,
         'IL6': 429,
         'IL8': 417,
         'BRAF': 305,
         'XYZB': 278,
         'n': 158,
         'KRAS': 148,
         'L': 144,
         'g': 140,
         'y': 140,
         'K': 140,
         'e': 140,
         'h': 140,
         't': 140,
         'NME1-NME2': 73,
         'COL1A1': 71,
         'UCHL1': 62,
         'NPIPA5': 58,
         'DSG1': 58,
         'MMP2': 57,
         'BASP1': 57,
         'FBN1': 55,
         'KRT14': 55,
         'MARCKS': 54,
         'CAV1': 54,
         'MAP2': 53,
         'DSP': 51,
         'COL6A1': 50,
         'TNS3': 50,
         'CXCL8': 49,
         'LRP1': 49,
         'COL3A1': 49,
         'KIF1A': 49,
         'RAP1GAP2': 49,
         'FN1': 48,
         'C1QA': 48,
         'RUNX1': 48,
         'PTPRC': 48,
         'COL4A1': 48,
         'NEDD4': 46,
         'FLT3': 46,
         'PROSER2': 46,
         'MAP1A': 46,
         'PPB

In [20]:
gene_counts["KRAS"]

148

### Load Ambiguous Gene Symbol Set

In [21]:
with open("../output/ambiguous_symbol_set.txt", "r") as file:
    # read each line, strip newline characters, and convert to a set
    ambiguous_symbol_set = set(line.strip() for line in file)
ambiguous_symbol_set

{'LINC00449',
 'NCA',
 'HOP',
 'FBXO31',
 'DSK2',
 'LEU2',
 'SWD2',
 'LAT1',
 'CAMC',
 'CAPS',
 'CD33L2',
 'COX',
 'MIG-2',
 'TDPX2',
 'BM600',
 'EIG7',
 'EK',
 'GDIL',
 'CARP',
 'PARF',
 'CGRP',
 'FRG',
 'DRB1',
 'DESI1',
 'HSP70-1',
 'LAD',
 'JP1',
 'SMA3',
 'MTTP',
 'ANCO1',
 'GUCA1A',
 'PDA2',
 'G15',
 'HCAP',
 'TLL',
 'THRRS',
 'TRS-AGA2-4',
 'MNAB',
 'FCHL2',
 'SYN',
 'AMPK',
 'UDPGT',
 'LINC01323',
 'ATXN8',
 'REH',
 'PKD4',
 'ZNF197-AS1',
 'CD32',
 'SMP1',
 'RNF51',
 'LST-2',
 'FAE',
 'EFL1',
 'HACL1',
 'MAR1',
 'P57',
 'CNAP1',
 'CT72',
 'BCDS1',
 'FCG2',
 'CPA3',
 'RBP5',
 'DAR',
 'AG2',
 'MYLK-AS2',
 'MPP7',
 'SHANK2-AS2',
 'NANOGP3',
 'CT45A2',
 'CSN1',
 'GP110',
 'SL1',
 'H2BFL',
 'DA1',
 'RAB1',
 'XPE',
 'SLC17A2',
 'RP7',
 'HPR1',
 'HGFL',
 'CMS1A',
 'AB14',
 'LINC02234',
 'PSGGA',
 'DP',
 'LILRA6',
 'P8',
 'J1',
 'MRE11B',
 'TPRKB',
 'P130',
 'POU5F1',
 'ACAT2',
 'CS1',
 'AE2',
 'LINC02815',
 'PACAP',
 'USE1',
 'TRYP2',
 'CT4.4',
 'FRMPD2',
 'TYRO4',
 'STRAP',
 'CERS1',

In [22]:
len(ambiguous_symbol_set)

5069

### How many of the interaction queries using genes included an ambiguous symbol?

In [23]:
data["ambiguous_gene_in_query"] = data.genes.apply(lambda x: x & ambiguous_symbol_set)
data_df = data[data.ambiguous_gene_in_query != set()]
data_df

time request_num  \
17     2023-04-03T21:41:46.386524      #112476   
18     2023-04-03T21:41:56.047208      #112476   
19     2023-04-03T21:41:57.140781      #112476   
20     2023-04-03T21:42:17.482316      #112476   
21     2023-04-03T21:42:27.478590      #112476   
...                            ...         ...   
23036   2023-04-16T01:17:14.363678      #37959   
23056   2023-04-16T01:17:19.774606      #37959   
23064   2023-04-16T01:17:35.989923      #37959   
23065   2023-04-16T01:17:36.045976      #37959   
23068   2023-04-16T01:17:37.637435      #37959   

                                       uuid method                       path  \
17     58e2f514-3c72-470c-9249-323d8c9bb081   POST  /api/v2/interactions.json   
18     fc4d3f38-9a45-4a1e-8572-ca7b789b34a4   POST  /api/v2/interactions.json   
19     88cb7915-5e7b-4807-9dfc-bdc77e8574f4   POST  /api/v2/interactions.json   
20     44602b18-ff6b-4c05-934e-71914d0080e6   POST  /api/v2/interactions.json   
21     6269711b-ec6b-4067-9d03-904711e64451   POST  /api/v2/interactions.json   
...                                     ...    ...                        ...   
23036  c83f3204-ff8a-46b9-976d-dfb0db6b9e62    GET  /api/v2/interactions.json   
23056  924bbe55-47a8-4b48-b0e5-8d06be1d4c87    GET  /api/v2/interactions.json   
23064  96b8aa02-0862-4b29-bba9-40934aa1caaa    GET  /api/v2/interactions.json   
23065  385d79d5-7991-4b4b-ad6e-aeb0886c77a0    GET  /api/v2/interactions.json   
23068  3e9c38bc-c066-46b6-a174-064702deb2a8    GET  /api/v2/interactions.json   

      format            controller        action  status  duration      view  \
17      json  ServicesV2Controller  interactions     200   9202.31   8841.23   
18      json  ServicesV2Controller  interactions     200   9223.01   8844.45   
19      json  ServicesV2Controller  interactions     200    213.81      1.15   
20      json  ServicesV2Controller  interactions     200   9401.45   9043.04   
21      json  ServicesV2Controller  interactions     200   9494.68   9129.97   
...      ...                   ...           ...     ...       ...       ...   
23036   json  ServicesV2Controller  interactions     200    516.33    433.78   
23056   json  ServicesV2Controller  interactions     200     50.15     11.97   
23064   json  ServicesV2Controller  interactions     200  15158.28  14676.73   
23065   json  ServicesV2Controller  interactions     200     53.55     13.20   
23068   json  ServicesV2Controller  interactions     200   1272.06   1162.75   

           db                                             params location  \
17     308.21  {'genes': 'TNF,AP1,AP2,XYZA', 'interaction_sou...      NaN   
18     312.74  {'genes': 'TNF,AP1,AP2,XYZA', 'interaction_sou...      NaN   
19     135.14  {'genes': 'TNF,AP1,AP2,XYZA', 'interaction_sou...      NaN   
20     305.14  {'genes': 'TNF,AP1,AP2,XYZA', 'interaction_sou...      NaN   
21     312.85  {'genes': 'TNF,AP1,AP2,XYZA', 'interaction_sou...      NaN   
...       ...                                                ...      ...   
23036   70.57        {'genes': 'RPL41P5,RPL13,PTPRR,PEBP1,MYL5'}      NaN   
23056   26.96        {'genes': 'TUBGCP2,RBM7,KMT2B,MRPL33,LDB1'}      NaN   
23064  439.61           {'genes': 'POLH,NPM1P2,MT-TL2,MSK9,MPI'}      NaN   
23065   28.44        {'genes': 'CLIP1,RN7SL5P,RBMY1C,PVR,QSOX1'}      NaN   
23068   94.18        {'genes': 'PTHLH,MTCP1,ITGAM,IDDM9,HSD3B2'}      NaN   

      error   type                                 genes  \
17      NaN  genes                 {AP2, XYZA, AP1, TNF}   
18      NaN  genes                 {AP2, XYZA, AP1, TNF}   
19      NaN  genes                 {AP2, XYZA, AP1, TNF}   
20      NaN  genes                 {AP2, XYZA, AP1, TNF}   
21      NaN  genes                 {AP2, XYZA, AP1, TNF}   
...     ...    ...                                   ...   
23036   NaN  genes  {PEBP1, MYL5, PTPRR, RPL13, RPL41P5}   
23056   NaN  genes  {LDB1, MRPL33, KMT2B, RBM7, TUBGCP2}   
23064   NaN  genes     {MT-TL2

In [24]:
num_interaction_queries_with_ambig_symbols = len(data_df)
num_interaction_queries_with_ambig_symbols

4033

### Of the total gene terms searched, how many were ambiguous gene symbols?

In [25]:
total_sum = 0

for key, value in gene_counts.items():
    if key in ambiguous_symbol_set:
        total_sum += value

In [26]:
num_queried_gene_terms_ambig = total_sum
num_queried_gene_terms_ambig

8570

### Of the unique gene symbols queried, how many are ambiguous?

In [27]:
filtered_dict = {key: value for key, value in gene_counts.items() if key in ambiguous_symbol_set}

ambiguous_symbol_queries_df = pd.DataFrame(
    list(filtered_dict.items()), columns=["Ambiguous Symbol", "# of Queries"]
)

ambiguous_symbol_queries_df = ambiguous_symbol_queries_df.sort_values(by="# of Queries", ascending=False)

ambiguous_symbol_queries_df.to_csv("../output/ambiguous_symbol_queries.csv", index=False)

In [28]:
num_unique_gene_symbols_queried_ambig = len(ambiguous_symbol_queries_df)
num_unique_gene_symbols_queried_ambig

1431

### Summary

In [29]:
title = [
    "# Interaction Queries using Genes",
    "# Terms as Genes for Queries",
    "# Unique Gene Symbols Queried",
]
summary_data = {
    "": [num_interaction_queries_using_genes, num_queried_gene_terms, num_unique_gene_symbols_queried],
    "# Ambiguous Symbols": [f"{num_interaction_queries_with_ambig_symbols} ({((num_interaction_queries_with_ambig_symbols/num_interaction_queries_using_genes)*100):.2f}%)", f"{num_queried_gene_terms_ambig} ({((num_queried_gene_terms_ambig/num_queried_gene_terms)*100):.2f}%)",f"{num_unique_gene_symbols_queried_ambig} ({((num_unique_gene_symbols_queried_ambig/num_unique_gene_symbols_queried)*100):.2f}%)"],
}
summary_df = pd.DataFrame(summary_data, index=title)
summary_df

# Ambiguous Symbols
# Interaction Queries using Genes   23070       4033 (17.48%)
# Terms as Genes for Queries       188832        8570 (4.54%)
# Unique Gene Symbols Queried       43400        1431 (3.30%)